xdsl/dialects/onnx.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+
+from typing import Annotated, TypeVar
+
+from xdsl.dialects.builtin import (
+    SSAValue,
+    TensorType,
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+)
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.utils.exceptions import VerifyException
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class ElementwiseBinOpBase(IRDLOperation):
+    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+    lhs = operand_def(TensorType[T])
+    rhs = operand_def(TensorType[T])
+    res = result_def(TensorType[T])
+    assembly_format = "`(` $lhs `,` $rhs `)` attr-dict `:` `(` type($lhs) `,` type($rhs) `)` `->` type($res)"
+
+    def __init__(
+        self,
+        lhs: SSAValue,
+        rhs: SSAValue,
+    ):
+        super().__init__(
+            operands=[lhs, rhs],
+            result_types=[[]],
+        )
```

Comment:
```suggestion
    def __init__(
        self,
        lhs: SSAValue,
        rhs: SSAValue,
        res_type: Attribute | None
    ):
        if res_type is None:
          res_type = lhs.type
        super().__init__(
            operands=[lhs, rhs],
            result_types=[res_type],
        )
```
The `res_type` argument is quite useful in the middle of a rewrite, when the operation would not verify.
Making it optional is quite useful as well, if by default you take the type of the first operand by example.

Reply:
Makes sense (my initial attempt was based on examples, which may not have been the most appropriate). I think by default the result type should be computed based on both operands, but in any case, the suggested modification is better.

Reply:
I think intiially, at least, I would not include `None` as an option to the result type because I think it's not essential to be able to omit.

Reply:
Sure, works as well!